File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # Time: O(n) ; Mem: O(1)
        res = max(nums)
        curr_min, curr_max = 1, 1

        for n in nums:
            temp_max = curr_max

            curr_max = max(n * curr_max, n * curr_min, n)
            curr_min = min(n * temp_max, n * curr_min, n)
            
            res = max(res, curr_max)
        
        return res

File: Data Structures & Algorithms/maximum-subarray/submission-2.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int res = nums[0];
        int curr = 0;

        for (int num : nums) {
            if (curr < 0)
                curr = 0;
            
            curr += num;
            res = max(res, curr);
        }
        
        return res;
    }
};